Cython/Build/Tests/TestRecythonize.py

```diff
@@ -0,0 +1,204 @@
+import shutil
+import os
+import tempfile
+
+import Cython.Build.Dependencies
+import Cython.Utils
+from Cython.TestUtils import CythonTest
+
+
+
+SAME = "The result of cytonization is the same"
+INCOR = "Incorrect cythonization"
+LINE_1 = '  /* "{name}{ext}":1\n'
+VARS_LINE = '  /*--- Wrapped vars code ---*/\n'
+
+
+class TestRecythonize(CythonTest):
+
+    def setUp(self):
+        CythonTest.setUp(self)
+        self.temp_dir = (
+            tempfile.mkdtemp(
+                prefix='recythonize-test',
+                dir='TEST_TMP' if os.path.isdir('TEST_TMP') else None
+            )
+        )
+        self.src_dir = tempfile.mkdtemp(prefix='src', dir=self.temp_dir)
+        self.dep_tree = Cython.Build.Dependencies.create_dependency_tree()
+        self.language_level = 3
+
+    def tearDown(self):
+        CythonTest.tearDown(self)
+        shutil.rmtree(self.temp_dir)
+
+    def fresh_cythonize(self, *args, **kwargs):
+        Cython.Utils.clear_function_caches()
+        Cython.Build.Dependencies._dep_tree = None  # discard method caches
+        kwargs.update(language_level=self.language_level)
+        Cython.Build.Dependencies.cythonize(*args, **kwargs)
+
+    def refresh_dep_tree(self):
+        Cython.Utils.clear_function_caches()
+        Cython.Utils.clear_method_caches(self.dep_tree)
+        self.dep_tree._transitive_cache.clear()
+
+    def fresh_all_dependencies(self, *args, **kwargs):
+        self.refresh_dep_tree()
+        return self.dep_tree.all_dependencies(*args, **kwargs)
+
+    def relative_lines(self, lines, line, start, end):
```

Comment:
I don't have any serious arguments against it, it's just a way to make it possible not to change the code if you need to have access to individual lines.
But if you think that way, in the context of this test case, such functionality will probably not be needed.

Reply:
In addition, by the name of the function, we should receive lines as an iterable, in the same format as the input `lines`.
Therefore, if you combine all this, then the names of the functions will become larger, that is, `relative_joined_lines`, `relative_joined_lines_from_file`.

Reply:
Or `read_code_from(start_code, offset, line_count)`.

Reply:
I solved the problem by adding `join` with the default: `relative_lines_from_file(self, path, line, start, end, join=True)`
I'm in the process of solving everything, then I will push all the commits at once so as not to strain CI.

Reply:
> Or `read_code_from(start_code, offset, line_count)`

Rather `read_lines_from(start_line, offset, line_count)` and `read_file` will store the lines in` self.last_read_lines` from where `self.read_code_from` will have access to them.

Quite confusing, even though we end up writing a little less code, I don't really like this approach.

Reply:
@scoder if you are satisfied with the implementation from [here](https://github.com/cython/cython/pull/4241/commits/a049908de6a1f023dd46a09eb462eb963a3ee2aa), then you can resolve the conversation

Reply:
> I solved the problem by adding `join` with the default: `relative_lines_from_file(self, path, line, start, end, join=True)`

YAGNI ?

Reply:
> YAGNI ?

`¯\_(ツ)_/¯` idk, but it looks like this

Reply:
Ok, @scoder is it enough to just make `relative_lines_from_file` always join lines?

Reply:
That's what I meant, yes. `relative_lines()` isn't used anywhere directly, is it?

Reply:
> `relative_lines()` isn't used anywhere directly, is it?

Because I just created them for this PR, they are not used anywhere except for the test case. In it, `relative_lines` was not used directly.
By the way, I think this (or `relative_lines_from_file`) function can be used to improve the output on failure in the similar [TestCyCache.py](https://github.com/cython/cython/blob/304b82267eda219fb801552341fa888b17f7db33/Cython/Build/Tests/TestCyCache.py#L52)

Reply:
If you are satisfied with the [implementation](https://github.com/cython/cython/pull/4241/commits/8b622df536a5fe1c40e632500c2b4ac61ce974a5), allow the conversation.

Reply:
If you are satisfied with the [implementation](https://github.com/cython/cython/pull/4241/commits/8b622df536a5fe1c40e632500c2b4ac61ce974a5), resolve the conversation.

Reply:
As I suggested before, I don't see a need for two methods. One seems enough. It can always be refactored later, if the need arises.

Reply:
resolved [here](https://github.com/cython/cython/pull/4241/commits/b6e795ef3ea50a5d797c5af2c7120df5d7b2eebc)